src/render/idyll-display/components/author-tool-buttons.js

```diff
@@ -0,0 +1,333 @@
+
+        if (chars.count === 0) {
+          break;
+        }
+      }
+    }
+  }
+
+  return range;
+}
+
+function setCurrentCursorPosition(node, chars) {
+  if (chars >= 0) {
+    var selection = window.getSelection();
+
+    const range = createRange(node.parentNode, { count: chars });
+
+    if (range) {
+      range.collapse(false);
+      selection.removeAllRanges();
+      selection.addRange(range);
+    }
+  }
+}
+
+class AuthorToolButtons extends React.PureComponent {
+  static contextType = Context;
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      showCode: false
+    };
+    this._markup = this.getMarkup(props);
+    cursorPositions[props.idyllASTNode.id] = -1;
+
+    this.domId = props.idyllASTNode.name + '-' + props.idyllASTNode.id;
+  }
+
+  updateNode(node, oldString, newString) {
+    let updated = false;
+    Object.keys(node.properties || {}).forEach(key => {
+      const prop = node.properties[key];
+      if (
+        prop.type === 'value' &&
+        ('' + prop.value).trim() == ('' + oldString).trim()
+      ) {
+        prop.value = newString;
+        updated = true;
+      }
+    });
+    let localUpdate = false;
+    (node.children || []).forEach(child => {
+      localUpdate = this.updateNode(child, oldString, newString);
+      updated = updated || localUpdate;
+    });
+    return updated;
+  }
+
+  // Flips between whether we are in the author view of a component
+  handleClickCode() {
+    if (this.state.showCode) {
+      const output = compile(this._markup, { async: false });
+      let node = output.children[0];
+      if (node.children && node.children.length) {
+        node = node.children[0];
+      }
+      const targetNode = getNodeById(
+        this.context.ast,
+        this.props.idyllASTNode.id
+      );
+      Object.keys(node).forEach(key => {
+        if (key === 'id') {
+          return;
+        }
+        targetNode[key] = node[key];
+      });
+
+      this.context.setAst(this.context.ast);
+      this.setState({
+        showCode: false
+      });
+    } else {
+      this._markup = this.getMarkup(this.props);
+      this.setState({ showCode: true });
+    }
+  }
+
+  handleClickProps() {
+    if (
+      this.context.activeComponent &&
+      this.context.activeComponent.id === this.props.idyllASTNode.id
+    ) {
+      this.context.setActiveComponent(null);
+    } else {
+      this.context.setActiveComponent({ ...this.props.idyllASTNode });
+    }
+  }
+
+  getMarkup(props) {
+    return AST.toMarkup({
+      id: -1,
+      type: 'component',
+      name: 'div',
+      children: [props.idyllASTNode]
+    });
+  }
+
+  registerObserver() {
+    if (!this._componentRef) {
+      return;
+    }
+    const node = getNodeById(this.context.ast, this.props.idyllASTNode.id);
+    let updated = false;
+    var observer = new MutationObserver(mutations => {
+      mutations.forEach(mutation => {
+        console.log('mutation', mutation);
```

Comment:
Do we want to keep this in the final merge?

Reply:
Will look at extracting this for next issue on code editing